### ISSUE TITLE
Matching generic return type of `SupportsGetItem`

### ISSUE THREAD

**Issue by cmp0xff** [astral-sh/ty#3988]
### Summary

# Describe the bug
`ty` does not match the generic return type of `SupportsGetItem` properly.

# Code or Screenshots

```python
from __future__ import annotations

from typing import TYPE_CHECKING, Any, TypeVar

if TYPE_CHECKING:
    from _typeshed import SupportsGetItem

class Series: ...

class DataFrame:
    def __getitem__(self, key: object) -> Series:
        return Series()

T = TypeVar("T")

def f(x: SupportsGetItem[Any, T]) -> T:
    if isinstance(x, DataFrame):
        return Series()
    else:
        return x[0]

```

# Error messages

```text
Return type does not match returned value: expected `T@f`, found `Series` (invalid-return-type) [Ln 18, Col 16]

    Expected `T@f` because of return type 
```

- `mypy` works fine: https://mypy-play.net/?gist=3aaea716bd3429a90a96457884f46712
- `ty` reports an error: https://play.ty.dev/61b04bb1-f388-4854-b0e6-ef6dfbd10f20

### Version

70e251023

**Comment by carljm** [comment]
Mypy is the only type checker which passes this example, and it does so for the wrong reasons entirely: it emits no diagnostic on `return Series()` because it considers `DataFrame` and `SupportsGetItem` incompatible, so it considers that branch unreachable. You can replace `return Series()` with `return 1` or `return "foo"` and it still won't error.

To pass this example for the right reasons requires a quite advanced feature, which I don't believe any Python type checker currently implements: tracking branch-local constraints on a typevar based on narrowing tests involving types generic over that typevar. ty matches the generic return type of `SupportsGetItem` just fine; we just don't infer from that a general constraint that `Series <: T` in that branch (and neither does any other type checker.)

We recently received another request for this same feature, and closed it as "unlikely to be implemented": #3868. Getting a second independent request is leading me to think maybe we should at least re-open that as a wish-level item.

**Comment by cmp0xff** [comment]
Thank you for the reply @carljm ! In my example above, if we test `reveal_type(f(DataFrame()))`, all of `mypy`, `pyright`, `pyrefly` and `ty` gives `Series`. Is this relevant?

**Comment by carljm** [comment]
@cmp0xff It is relevant to demonstrating that your function is in fact safe; unfortunately it's not very relevant to the additional work that a type checker needs to do in order to prove that it's safe.
